### DISCUSSION TITLE
Creating a Vector/Multivariate forecaster using `make_reduction`

### DISCUSSION THREAD

**Discussion by mabilton** [alan-turing-institute/sktime#3692] [Q&A]
Hi there.

This question is similar in nature to #1828, albeit with a slightly different focus: is it possible to use `make_reduction` to create a vector/multivariate forecaster? 

From my understanding, when a `make_reduction` forecaster is given a multivariate time series with `N` columns, it simply produces a univariate forecaster for each column (i.e. `N` independent forecasters are fitted), where each forecaster is trained exclusively on the (lagged) features of the column it is forecasting.

With this in mind, is it possible to create a vector forecaster by training, say, a `MultiTaskLasso` from `sklearn` on the lagged features of all `N` columns and then predicting all `N` features simultaneously? 

Thanks in advance for any help.

Cheers,
Matt.

**Comment by fkiraly** [comment]
Yes, this is possible if you call the `DirectReductionForecaster` directly.

Currently, it is a prototype alternative and not accessible via `make_reduction`, but you can obtain it via

```from sktime.forecasting.compose._reduce import DirectReductionForecaster```

In the future, this feature may be able via `make_reduction` directly.

**Reply by mabilton** [reply]
Hi Franz.

Thanks for your timely response. Admittedly, I may have missed something, but it appears that calling `DirectReductionForecaster` directly still results in fitting individual univariate models to each column. Just as a quick MWE to illustrate what I mean:
```python
import warnings
warnings.filterwarnings('ignore')
from sklearn.linear_model import MultiTaskLasso 
from sktime.forecasting.compose._reduce import DirectReductionForecaster
from sktime.datasets import load_macroeconomic
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.base import ForecastingHorizon

# max_iter=1 to minimise train time
forecaster = DirectReductionForecaster(MultiTaskLasso(max_iter=1))
y = load_macroeconomic()
y_train, y_test = temporal_train_test_split(y, test_size=36)
fh = ForecastingHorizon(y_train.index, is_relative=False)
forecaster.fit(y_train, fh=fh)
# forecaster appears to comprise of many univariate models:
forecaster.get_fitted_params()['forecasters']
```
Shouldn’t I expect `forecaster.get_fitted_params()` to contain a single forecaster (i.e. a single model which accepts the lagged features of all `N` columns and returns `N*len(fh)` predictions)? 

As an aside, I’ve also had a chance to look through the underlying code for `make_reduction` and, based upon what I’ve seen, it appears that the answer to my initial question is ‘No’. More specifically, the `fit` method of the `BaseForecaster` class (which `DirectReductionForecaster` also inherits from) automatically ‘vectorises’ over the columns of `y` if it is a multi-column dataframe (see [here](https://github.com/sktime/sktime/blob/18b502e969d9c8d8dafce5cea78eb0c43fe23d91/sktime/forecasting/base/_base.py#L297)). It seems then that the only way to achieve what I’m looking to do would be to modify the `fit` method of the `BaseForecaster`? Once again, apologies if I’ve missed something or misinterpreted your response.

**Reply by mabilton** [reply]
Just another quick update from me: after digging a bit deeper, it appears that all of the classes in the `_reduce` module inherit their `'scitype:y'` tag from the `BaseForecaster` class, which is set to `'univariate'` by default; an apparent implication of this is that any forecaster produced using these `_reduce` classes will automatically be vectorised over each column of a multivariate time series. As a naive solution, I first tried to manually change the `'scitype:y'` tag from `'univariate'` to `'multivariate'`, like so:
```python
import warnings
warnings.filterwarnings('ignore')
from sklearn.linear_model import MultiTaskLasso 
from sktime.forecasting.compose._reduce import DirectReductionForecaster
from sktime.datasets import load_macroeconomic
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.base import ForecastingHorizon

y = load_macroeconomic()
y_train, y_test = temporal_train_test_split(y, test_size=1)
fh = ForecastingHorizon([1], is_relative=True)
forecaster = DirectReductionForecaster(MultiTaskLasso(), window_length=2)
# Manually set 'scitype:y' tag from 'univariate' to 'multivariate':
forecaster = forecaster.set_tags(**{'scitype:y': 'multivariate'})
print(f"Tags before fitting: {forecaster.get_tags()}") # Shows  'scitype:y' tag as 'multivariate'
forecaster.fit(y_train, fh=fh)
print(f"Tags after fitting: {forecaster.get_tags()}") # Shows  'scitype:y' tag as 'univariate'
# Still have one Lasso model fitted to each column:
forecaster.forecasters_
```
Unfortunately, this still doesn't get the job done since the `reset` method is called at the start of the `fit` method of the `BaseForecaster` class (from which `DirectReductionForecaster` inherits), which changes back the `scitype:y` tag back to `'univariate'`. As a 'hackish' way to get around this, I overrode the `reset` method so that this tag isn't reset at the start of the `fit` method, like so:
```python
import warnings
warnings.filterwarnings('ignore')
from sklearn.linear_model import MultiTaskLasso 
from sktime.forecasting.compose._reduce import DirectReductionForecaster
from sktime.datasets import load_macroeconomic
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.base import ForecastingHorizon

y = load_macroeconomic()
y_train, y_test = temporal_train_test_split(y, test_size=1)
fh = ForecastingHorizon([1], is_relative=True)
forecaster = DirectReductionForecaster(MultiTaskLasso(), window_length=2)
# Manually set 'scitype:y' tag from 'univariate' to 'multivariate':
forecaster = forecaster.set_tags(**{'scitype:y': 'multivariate'})
# Override reset method so that 'scitype:y' tag is not changed from 'multivariate' to 'univariate' upon calling fit:
forecaster.reset = lambda : None
forecaster.fit(y_train, fh=fh)
# Now only a single Lasso model is fit:
print(forecaster.estimators_)
# The shape of the coefficients is (12, 24), as expected
# (i.e. we're predicting num_features*len(fh) = 12*1 = 12 labels from num_features*window_length = 12*2 = 24 features
print(forecaster.estimators_[0].coef_.shape)
``` 
Although this does appear to work, it's obviously not an ideal solution - I'll look to open an issue ticket about this either later today or tomorrow. In any case, thanks for the help Franz.

**Comment by fkiraly** [comment]
@MABilton, you need to use `X_treatment="shifted"`, only then does it make mathematical/algorithmic sense to use a multi-output regressor. Otherwise the `X` you predict from will be different per column of `y`.

See description of the what to predict and from what, in the docstring.

Although I might be wrong, happy to be convinced otherwise

**Comment by fkiraly** [comment]
Waittttttt - I think I have misunderstood you!

You want the multi-output regressor to be applied *to all variables* rather than to be applied *to multiple horizon indices for the same variable*!

That is currently not possible, but I was working on it here:
https://github.com/sktime/sktime/pull/3382

**Reply by mabilton** [reply]
Hi Franz.

That's correct - apologies for being 'fast and loose' with my terminology, I'm somewhat new to the time series literature. Just to make sure we're all on the same page (and also to help me clarify my thoughts), I think it'd pay for me to be a bit more formal about my problem description:

Suppose we have observations from $k$ separate time series; I shall use $y_i(t)$ to denote the (scalar) value from the $i$ th time series observed at time $t$. Similarly, I'll use $Y(t) = [y_1(t), ..., y_k(t)]^T$ to denote all of the observations made at time $t$ 'stacked' together as a column vector. Assuming we only want to predict one step into the future (as is the case with the 'recursive' strategy, although the discussion here should generalise quite easily to the 'direct' and 'dirrec' cases), the model we wish to fit is of the form:
$$Y(t) = c +  \sum_{i=1}^{L} A_{i} Y(t-i) + e(t) = \big[ c^T, A_1, ..., A_L \big]  \big[ \vec{1}^T, Y(t-1)^T, ..., Y(t-L)^T \big]^T + e(t) = Ax(t) + e(t)$$

where $c$ is a $k$-dimensional intercept vector, $L$ is the maximum lag we want to consider, the $A_i$'s are $(k, k)$-shaped coefficient matrices, $e(t)$ is a $k$-dimensional noise vector, $A$ is a $\big(k, k \cdot (L + 1) \big)$-shaped matrix of 'grouped' coefficients, and $x(t)$ is a $\big(k \cdot (L+1)\big)$-dimensional feature vector where all of the lagged features have been stacked together into a single column. Naturally, if we've observed $m$ iid observations of $Y(t)$ and $x(t)$ occurring at times $t \in \\{t_1, ..., t_m\\}$, we can stack all these observations together as rows into a single matrix to form the following system of equations:
$$\big[ Y(t_1), ..., Y(t_m) \big]^T = Y =  \big[A x(t_1), ..., A x(t_m) \big]^T +  [e(t_1), ..., e(t_m)]^T =  \big[x(t_t), ..., x(t_m) \big]^T A^T + [e(t_1), ..., e(t_m)]^T = X A^T + E$$

or, more tersely:
$$\implies Y = X A^T + E$$ 

Here, $Y$ is a $(m, k)$-shaped target matrix, $X$ is a $(m, k \cdot (L+1))$-shaped feature matrix, and $E$ is a $(m,k)$-shaped noise matrix. This problem has the least-squares solution:
$$\hat{A}^T = (X^T X)^{-1} X^T Y$$

So in summary, I'm basically trying to fit a Vector Autoregression Model (VAR), but instead of computing the least-squares solution for $\hat{A}$, I'd like to use L1-regularised least squares (aka Lasso regression) to compute $\hat{A}$. With that being said, this problem set-up also naturally generalises to the case of using an arbitrary, potentially non-linear `sklearn` estimator to predict $Y(t)$, i.e.:
$$Y(t) = f \big( Y(t-1), ..., Y(t-L) \big) +  e(t) $$

where $f$ is a potentially non-linear estimator.

From what I've observed, it appears that when presented with $k$-dimensional time series data (i.e. a data frame with $m$ observation rows and $k$ feature columns), the `DirectReductionForecaster` is fitting the model:
$$y_j(t) = c_j +  \sum_{i=1}^{L} a_{ij} y_j(t-i) + e_j(t) $$
to each column individually for $j \in \\{1, ..., k \\}$ (i.e. to predict the $j$ th time series, `DirectReductionForecaster` is only looking at the lagged features of the $j$ th time series).

Hopefully this write-up has made it clear what I'm trying to achieve; apologies in advanced if my notation is unclear or if there's an obvious mistake I've missed. If this is the kind of functionality #3382 is looking to add to sktime, I'd be happy to help contribute to that once I've read over the contribution guides. Thanks for all your help.
